server/android_notification_server.go

```diff
@@ -240,6 +234,73 @@ func (me *AndroidNotificationServer) SendNotification(msg *PushNotification) Pus
 	return NewOkPushResponse()
 }
 
+func (me *AndroidNotificationServer) SendNotificationWithRetry(fcmMsg *messaging.Message) error {
+	var err error
+	waitTime := time.Second
+
+	logger := me.logger.With(mlog.String("did", fcmMsg.Token))
+
+	// Keep a general context to make sure the whole retry
+	// doesn't take longer than the timeout.
+	generalContext, cancelGeneralContext := context.WithTimeout(context.Background(), me.sendTimeout)
+	defer cancelGeneralContext()
+
+	for retries := 0; retries < MAX_RETRIES; retries++ {
+		start := time.Now()
+
+		retryContext, cancelRetryContext := context.WithTimeout(generalContext, me.retryTimeout)
+		defer cancelRetryContext()
```

Comment:
nit: these cancellations queue up to run at when `SendNotificationWithRetry` returns. I don't think that causes any issues, but a `defer` in a `for` loop always gives me pause in case the semantics aren't understood by a future reader. Is it worth considering calling this explicitly at the end of the loop, or wrapping in a closure?

Reply:
I am 0/5 on this. I feel adding extra boilerplate will just cause more cognitive overload, instead of clarify things.

You are right that having these defer here are a bit counterintuitive (they are not called at the end of the loop, but when the function finishes) but we would need to cancel then on every break, which is also quite boilerplaty. I feel it is OK like this, but I would be happy to hear other options.